docs/api.rst

```diff
@@ -1,44 +1,117 @@
+      return web.Application(
+         [
+            # add your handlers here
+         ], **settings)
 
-   If any callback raises an exception, then the application is
-   terminated **before** the IOLoop is started.
+   if __name__ == '__main__':
+      sprockets.http.run(create_application)
 
-:on_start:
-   This set of callbacks is invoked after Tornado forks sub-processes
-   (based on the ``number_of_procs`` setting) and **after**
-   :meth:`~tornado.ioloop.IOLoop.start` is called.
+Since :func:`sprockets.http.run` accepts any callable, you can pass a
+class instance in as well.  The :class:`sprockets.http.app.Application`
+is a specialization of :class:`tornado.web.Application` that includes
+state management callbacks that work together with the ``run`` function
+and provide hooks for performing initialization and shutdown tasks.
 
-:shutdown:
-   When the application receives a stop signal, it will run each of the
-   callbacks before terminating the application instance.  Exceptions
-   raised by the callbacks are simply logged.
+The following example uses :class:`sprockets.http.app.Application` as a
+base class to implement asynchronously connecting to a mythical database
+when the application starts.
 
-See :func:`sprockets.http.run` for a detailed description of how to
-install the runner callbacks.
+.. code-block:: python
+   :caption: Using the Application class
+
+   from tornado import locks, web
+   from sprockets.http import app, run
+
+   class Application(app.Application):
+      def __init__(self, *args, **kwargs):
+         handlers = [
+            # insert your handlers here
+         ]
+         super(Application, self).__init__(handlers, *args, **kwargs)
+         self.ready_to_serve = locks.Event()
+         self.ready_to_serve.clear()
+         self.on_start_callbacks.append(self._connect_to_database)
+
+      def _connect_to_database(self, _self, iol):
+         def on_connected(future):
+            if future.exception():
+               iol.call_later(0.5, self._connect_to_database, _self, iol)
+            else:
+               self.ready_to_serve.set()
+
+         future = dbconnector.connect()
+         iol.add_future(future, on_connected)
+
+   if __name__ == '__main__':
+      run(Application)
+
+Implementing a ``ready_to_serve`` event is a useful paradigm for applications
+that need to asynchronously initialize before they can service requests.  We
+can continue the example and add a ``/status`` endpoint that makes use of
+the event:
 
-Internal Interfaces
-~~~~~~~~~~~~~~~~~~~
-.. automodule:: sprockets.http.runner
-   :members:
+.. code-block:: python
+   :caption: Implementing health checks
+
+   class StatusHandler(web.RequestHandler):
+      @gen.coroutine
+      def prepare(self):
+         maybe_future = super(StatusHandler, self).prepare()
+         if concurrent.is_future(maybe_future):
+            yield maybe_future
+         if not self._finished and not self.ready_to_serve.is_set():
```

Comment:
Perhaps a nit, but it's not obvious where ``self.ready_to_serve`` is defined. Perhaps, this should be ``self.application.ready_to_serve.is_set()``?